### ISSUE TITLE
incorporate upstream fixes to crc32c code

### ISSUE THREAD

**Issue by stevengj** [JuliaLang/julia#52325]
Our [CRC32c implementation](https://github.com/JuliaLang/julia/blob/5b2fcb68800875e570d7bb8c78ed00d360b6cfd5/src/crc32c.c) was adapted from @madler's (BSD-licensed) code in [this stackoverflow question](https://stackoverflow.com/questions/17645167).   However, I just noticed that he posted a [2021 comment](https://stackoverflow.com/questions/17645167/implementing-sse-4-2s-crc32c-in-software#comment119798529_17646775) that the code has been updated from the version we used:

> SGJ I have updated the code in this answer. The register constraints in the assembly code, which was inherited in your code, had a problem which is now fixed. As well as some other improvements.

I don't know if there is an easy way to get the diff, but we should probably go through it and incorporate any improvements as needed.

For reference, [this](https://github.com/JuliaLang/julia/blob/155437a4b2aa7f31988e4007ced5f5abe22812ad/src/support/crc32c.c) is the version of `crc32c.c` from my initial import, which should be closest to Mark's original code.

*Update*: the `diff` is available on stackoverflow here: https://stackoverflow.com/posts/17646775/revisions … note that there were 3 updates in 2021.

**Comment by stevengj** [comment]
StackOverflow's diff format is unreadable, so here are the 2021 changes via `diff -u`:
```diff
--- old.md	2023-11-27 18:01:42
+++ new.md	2023-11-27 18:02:32
@@ -4,11 +4,15 @@
 
 
 
+crc32c.c:
+
+
+
     /* crc32c.c -- compute CRC-32C using the Intel crc32 instruction
 
-     * Copyright (C) 2013 Mark Adler
+     * Copyright (C) 2013, 2021 Mark Adler
 
-     * Version 1.1  1 Aug 2013  Mark Adler
+     * Version 1.2  5 Jun 2021  Mark Adler
 
      */
 
@@ -56,706 +60,610 @@
 
 
 
-    /* Use hardware CRC instruction on Intel SSE 4.2 processors.  This computes a
+    /* Version History:
 
-       CRC-32C, *not* the CRC-32 used by Ethernet and zip, gzip, etc.  A software
+     1.0  10 Feb 2013  First version
 
-       version is provided as a fall-back, as well as for speed comparisons. */
+     1.1  31 May 2021  Correct register constraints on assembly instructions
 
+                       Include pre-computed tables to avoid use of pthreads
 
+                       Return zero for the CRC when buf is NULL, as initial value
 
-    /* Version history:
+     1.2   5 Jun 2021  Make tables constant
 
-       1.0  10 Feb 2013  First version
+     */
 
-       1.1   1 Aug 2013  Correct comments on why three crc instructions in parallel
 
-     */
 
+    // Use hardware CRC instruction on Intel SSE 4.2 processors.  This computes a
 
+    // CRC-32C, *not* the CRC-32 used by Ethernet and zip, gzip, etc.  A software
 
-    #include <stdio.h>
+    // version is provided as a fall-back, as well as for speed comparisons.
 
-    #include <stdlib.h>
 
-    #include <stdint.h>
 
-    #include <unistd.h>
+    #include <stddef.h>
 
-    #include <pthread.h>
+    #include <stdint.h>
 
 
 
-    /* CRC-32C (iSCSI) polynomial in reversed bit order. */
+    // Tables for CRC word-wise calculation, definitions of LONG and SHORT, and CRC
 
-    #define POLY 0x82f63b78
+    // shifts by LONG and SHORT bytes.
 
+    #include "crc32c.h"
 
 
-    /* Table for a quadword-at-a-time software crc. */
 
-    static pthread_once_t crc32c_once_sw = PTHREAD_ONCE_INIT;
+    // Table-driven software version as a fall-back.  This is about 15 times slower
 
-    static uint32_t crc32c_table[8][256];
+    // than using the hardware instructions.  This assumes little-endian integers,
 
+    // as is the case on Intel processors that the assembler code here is for.
 
+    static uint32_t crc32c_sw(uint32_t crc, void const *buf, size_t len) {
 
-    /* Construct table for software CRC-32C calculation. */
+        if (buf == NULL)
 
-    static void crc32c_init_sw(void)
+            return 0;
 
-    {
+        unsigned char const *data = buf;
 
-        uint32_t n, crc, k;
+        while (len && ((uintptr_t)data & 7) != 0) {
 
+            crc = (crc >> 8) ^ crc32c_table[0][(crc ^ *data++) & 0xff];
 
+            len--;
 
-        for (n = 0; n < 256; n++) {
+        }
 
-            crc = n;
+        size_t n = len >> 3;
 
-            crc = crc & 1 ? (crc >> 1) ^ POLY : crc >> 1;
+        for (size_t i = 0; i < n; i++) {
 
-            crc = crc & 1 ? (crc >> 1) ^ POLY : crc >> 1;
+            uint64_t word = crc ^ ((uint64_t const *)data)[i];
 
-            crc = crc & 1 ? (crc >> 1) ^ POLY : crc >> 1;
+            crc = crc32c_table[7][word & 0xff] ^
 
-            crc = crc & 1 ? (crc >> 1) ^ POLY : crc >> 1;
+                  crc32c_table[6][(word >> 8) & 0xff] ^
 
-            crc = crc & 1 ? (crc >> 1) ^ POLY : crc >> 1;
+                  crc32c_table[5][(word >> 16) & 0xff] ^
 
-            crc = crc & 1 ? (crc >> 1) ^ POLY : crc >> 1;
+                  crc32c_table[4][(word >> 24) & 0xff] ^
 
-            crc = crc & 1 ? (crc >> 1) ^ POLY : crc >> 1;
+                  crc32c_table[3][(word >> 32) & 0xff] ^
 
-            crc = crc & 1 ? (crc >> 1) ^ POLY : crc >> 1;
+                  crc32c_table[2][(word >> 40) & 0xff] ^
 
-            crc32c_table[0][n] = crc;
+                  crc32c_table[1][(word >> 48) & 0xff] ^
 
+                  crc32c_table[0][word >> 56];
+
         }
 
-        for (n = 0; n < 256; n++) {
+        data += n << 3;
 
-            crc = crc32c_table[0][n];
+        len &= 7;
 
-            for (k = 1; k < 8; k++) {
+        while (len) {
 
-                crc = crc32c_table[0][crc & 0xff] ^ (crc >> 8);
+            len--;
 
-                crc32c_table[k][n] = crc;
+            crc = (crc >> 8) ^ crc32c_table[0][(crc ^ *data++) & 0xff];
 
-            }
-
         }
 
+        return crc;
+
     }
 
 
 
-    /* Table-driven software version as a fall-back.  This is about 15 times slower
+    // Apply the zeros operator table to crc.
 
-       than using the hardware instructions.  This assumes little-endian integers,
+    static uint32_t crc32c_shift(uint32_t const zeros[][256], uint32_t crc) {
 
-       as is the case on Intel processors that the assembler code here is for. */
+        return zeros[0][crc & 0xff] ^ zeros[1][(crc >> 8) & 0xff] ^
 
-    static uint32_t crc32c_sw(uint32_t crci, const void *buf, size_t len)
+               zeros[2][(crc >> 16) & 0xff] ^ zeros[3][crc >> 24];
 
-    {
+    }
 
-        const unsigned char *next = buf;
 
-        uint64_t crc;
 
+    // Compute CRC-32C using the Intel hardware instruction. Three crc32q
 
+    // instructions are run in parallel on a single core. This gives a
 
-        pthread_once(&crc32c_once_sw, crc32c_init_sw);
+    // factor-of-three speedup over a single crc32q instruction, since the
 
-        crc = crci ^ 0xffffffff;
+    // throughput of that instruction is one cycle, but the latency is three
 
-        while (len && ((uintptr_t)next & 7) != 0) {
+    // cycles.
 
-            crc = crc32c_table[0][(crc ^ *next++) & 0xff] ^ (crc >> 8);
+    static uint32_t crc32c_hw(uint32_t crc, void const *buf, size_t len) {
 
-            len--;
+        if (buf == NULL)
 
-        }
+            return 0;
 
-        while (len >= 8) {
 
-            crc ^= *(uint64_t *)next;
 
-            crc = crc32c_table[7][crc & 0xff] ^
+        // Pre-process the crc.
 
-                  crc32c_table[6][(crc >> 8) & 0xff] ^
+        uint64_t crc0 = crc ^ 0xffffffff;
 
-                  crc32c_table[5][(crc >> 16) & 0xff] ^
 
-                  crc32c_table[4][(crc >> 24) & 0xff] ^
 
-                  crc32c_table[3][(crc >> 32) & 0xff] ^
+        // Compute the crc for up to seven leading bytes, bringing the data pointer
 
-                  crc32c_table[2][(crc >> 40) & 0xff] ^
+        // to an eight-byte boundary.
 
-                  crc32c_table[1][(crc >> 48) & 0xff] ^
+        unsigned char const *next = buf;
 
-                  crc32c_table[0][crc >> 56];
+        while (len && ((uintptr_t)next & 7) != 0) {
 
-            next += 8;
+            __asm__("crc32b\t" "(%1), %0"
 
-            len -= 8;
+                    : "+r"(crc0)
 
-        }
+                    : "r"(next), "m"(*next));
 
-        while (len) {
+            next++;
 
-            crc = crc32c_table[0][(crc ^ *next++) & 0xff] ^ (crc >> 8);
-
             len--;
 
         }
 
-        return (uint32_t)crc ^ 0xffffffff;
 
-    }
 
+        // Compute the crc on sets of LONG*3 bytes, making use of three ALUs in
 
+        // parallel on a single core.
 
-    /* Multiply a matrix times a vector over the Galois field of two elements,
+        while (len >= LONG*3) {
 
-       GF(2).  Each element is a bit in an unsigned integer.  mat must have at
+            uint64_t crc1 = 0;
 
-       least as many entries as the power of two for most significant one bit in
+            uint64_t crc2 = 0;
 
-       vec. */
+            unsigned char const *end = next + LONG;
 
-    static inline uint32_t gf2_matrix_times(uint32_t *mat, uint32_t vec)
+            do {
 
-    {
+                __asm__("crc32q\t" "(%3), %0\n\t"
 
-        uint32_t sum;
+                        "crc32q\t" LONGx1 "(%3), %1\n\t"
 
+                        "crc32q\t" LONGx2 "(%3), %2"
 
+                        : "+r"(crc0), "+r"(crc1), "+r"(crc2)
 
-        sum = 0;
+                        : "r"(next), "m"(*next));
 
-        while (vec) {
+                next += 8;
 
-            if (vec & 1)
+            } while (next < end);
 
-                sum ^= *mat;
+            crc0 = crc32c_shift(crc32c_long, crc0) ^ crc1;
 
-            vec >>= 1;
+            crc0 = crc32c_shift(crc32c_long, crc0) ^ crc2;
 
-            mat++;
+            next += LONG*2;
 
+            len -= LONG*3;
+
         }
 
-        return sum;
 
-    }
 
+        // Do the same thing, but now on SHORT*3 blocks for the remaining data less
 
+        // than a LONG*3 block.
 
-    /* Multiply a matrix by itself over GF(2).  Both mat and square must have 32
+        while (len >= SHORT*3) {
 
-       rows. */
+            uint64_t crc1 = 0;
 
-    static inline void gf2_matrix_square(uint32_t *square, uint32_t *mat)
+            uint64_t crc2 = 0;
 
-    {
+            unsigned char const *end = next + SHORT;
 
-        int n;
+            do {
 
+                __asm__("crc32q\t" "(%3), %0\n\t"
 
+                        "crc32q\t" SHORTx1 "(%3), %1\n\t"
 
-        for (n = 0; n < 32; n++)
+                        "crc32q\t" SHORTx2 "(%3), %2"
 
-            square[n] = gf2_matrix_times(mat, mat[n]);
+                        : "+r"(crc0), "+r"(crc1), "+r"(crc2)
 
-    }
+                        : "r"(next), "m"(*next));
 
+                next += 8;
 
+            } while (next < end);
 
-    /* Construct an operator to apply len zeros to a crc.  len must be a power of
+            crc0 = crc32c_shift(crc32c_short, crc0) ^ crc1;
 
-       two.  If len is not a power of two, then the result is the same as for the
+            crc0 = crc32c_shift(crc32c_short, crc0) ^ crc2;
 
-       largest power of two less than len.  The result for len == 0 is the same as
+            next += SHORT*2;
 
-       for len == 1.  A version of this routine could be easily written for any
+            len -= SHORT*3;
 
-       len, but that is not needed for this application. */
+        }
 
-    static void crc32c_zeros_op(uint32_t *even, size_t len)
 
-    {
 
-        int n;
+        // Compute the crc on the remaining eight-byte units less than a SHORT*3
 
-        uint32_t row;
+        // block.
 
-        uint32_t odd[32];       /* odd-power-of-two zeros operator */
+        unsigned char const *end = next + (len - (len & 7));
 
+        while (next < end) {
 
+            __asm__("crc32q\t" "(%1), %0"
 
-        /* put operator for one zero bit in odd */
+                    : "+r"(crc0)
 
-        odd[0] = POLY;              /* CRC-32C polynomial */
+                    : "r"(next), "m"(*next));
 
-        row = 1;
+            next += 8;
 
-        for (n = 1; n < 32; n++) {
-
-            odd[n] = row;
-
-            row <<= 1;
-
         }
 
+        len &= 7;
 
 
-        /* put operator for two zero bits in even */
 
-        gf2_matrix_square(even, odd);
+        // Compute the crc for up to seven trailing bytes.
 
+        while (len) {
 
+            __asm__("crc32b\t" "(%1), %0"
 
-        /* put operator for four zero bits in odd */
+                    : "+r"(crc0)
 
-        gf2_matrix_square(odd, even);
+                    : "r"(next), "m"(*next));
 
+            next++;
 
-
-        /* first square will put the operator for one zero byte (eight zero bits),
-
-           in even -- next square puts operator for two zero bytes in odd, and so
+            len--;
 
-           on, until len has been rotated down to zero */
+        }
 
-        do {
 
-            gf2_matrix_square(even, odd);
 
-            len >>= 1;
+        // Return the crc, post-processed.
 
-            if (len == 0)
+        return ~(uint32_t)crc0;
 
-                return;
+    }
 
-            gf2_matrix_square(odd, even);
 
-            len >>= 1;
 
-        } while (len);
+    // Check for SSE 4.2.  SSE 4.2 was first supported in Nehalem processors
 
+    // introduced in November, 2008.  This does not check for the existence of the
 
+    // cpuid instruction itself, which was introduced on the 486SL in 1992, so this
 
-        /* answer ended up in odd -- copy to even */
+    // will fail on earlier x86 processors.  cpuid works on all Pentium and later
 
-        for (n = 0; n < 32; n++)
+    // processors.
 
-            even[n] = odd[n];
+    #define SSE42(have) \
 
-    }
+        do { \
 
+            uint32_t eax, ecx; \
 
+            eax = 1; \
 
-    /* Take a length and build four lookup tables for applying the zeros operator
+            __asm__("cpuid" \
 
-       for that length, byte-by-byte on the operand. */
+                    : "=c"(ecx) \
 
-    static void crc32c_zeros(uint32_t zeros[][256], size_t len)
+                    : "a"(eax) \
 
-    {
+                    : "%ebx", "%edx"); \
 
-        uint32_t n;
+            (have) = (ecx >> 20) & 1; \
 
-        uint32_t op[32];
+        } while (0)
 
 
 
-        crc32c_zeros_op(op, len);
+    // Compute a CRC-32C.  If the crc32 instruction is available, use the hardware
 
-        for (n = 0; n < 256; n++) {
+    // version.  Otherwise, use the software version.
 
-            zeros[0][n] = gf2_matrix_times(op, n);
+    uint32_t crc32c(uint32_t crc, void const *buf, size_t len) {
 
-            zeros[1][n] = gf2_matrix_times(op, n << 8);
+        int sse42;
 
-            zeros[2][n] = gf2_matrix_times(op, n << 16);
+        SSE42(sse42);
 
-            zeros[3][n] = gf2_matrix_times(op, n << 24);
+        return sse42 ? crc32c_hw(crc, buf, len) : crc32c_sw(crc, buf, len);
 
-        }
-
     }
 
 
 
-    /* Apply the zeros operator table to crc. */
+Code to generate crc32c.h (stackoverflow won't let me post the tables themselves, due to a 30,000 character limit in an answer):
 
-    static inline uint32_t crc32c_shift(uint32_t zeros[][256], uint32_t crc)
 
-    {
 
-        return zeros[0][crc & 0xff] ^ zeros[1][(crc >> 8) & 0xff] ^
+    // Generate crc32c.h for crc32c.c.
 
-               zeros[2][(crc >> 16) & 0xff] ^ zeros[3][crc >> 24];
 
-    }
 
+    #include <stdio.h>
 
+    #include <stdint.h>
 
-    /* Block sizes for three-way parallel crc computation.  LONG and SHORT must
 
-       both be powers of two.  The associated string constants must be set
 
-       accordingly, for use in constructing the assembler instructions. */
-
     #define LONG 8192
 
-    #define LONGx1 "8192"
-
-    #define LONGx2 "16384"
-
     #define SHORT 256
 
-    #define SHORTx1 "256"
 
-    #define SHORTx2 "512"
 
+    // Print a 2-D table of four-byte constants in hex.
 
+    static void print_table(uint32_t *tab, size_t rows, size_t cols, char *name) {
 
-    /* Tables for hardware crc that shift a crc by LONG and SHORT zeros. */
+        printf("static uint32_t const %s[][%zu] = {\n", name, cols);
 
-    static pthread_once_t crc32c_once_hw = PTHREAD_ONCE_INIT;
+        size_t end = rows * cols;
 
-    static uint32_t crc32c_long[4][256];
+        size_t k = 0;
 
-    static uint32_t crc32c_short[4][256];
+        for (;;) {
 
+            fputs("   {", stdout);
 
+            size_t n = 0, j = 0;
 
-    /* Initialize tables for shifting crcs. */
+            for (;;) {
 
-    static void crc32c_init_hw(void)
+                printf("0x%08x", tab[k + n]);
 
-    {
+                if (++n == cols)
 
-        crc32c_zeros(crc32c_long, LONG);
+                    break;
 
-        crc32c_zeros(crc32c_short, SHORT);
+                putchar(',');
 
-    }
+                if (++j == 6) {
 
+                    fputs("\n   ", stdout);
 
+                    j = 0;
 
-    /* Compute CRC-32C using the Intel hardware instruction. */
+                }
 
-    static uint32_t crc32c_hw(uint32_t crc, const void *buf, size_t len)
+                putchar(' ');
 
-    {
+            }
 
-        const unsigned char *next = buf;
+            k += cols;
 
-        const unsigned char *end;
+            if (k == end)
 
-        uint64_t crc0, crc1, crc2;      /* need to be 64 bits for crc32q */
+                break;
 
+            puts("},");
 
+        }
 
-        /* populate shift tables the first time through */
+        puts("}\n};");
 
-        pthread_once(&crc32c_once_hw, crc32c_init_hw);
+    }
 
 
 
-        /* pre-process the crc */
+    /* CRC-32C (iSCSI) polynomial in reversed bit order. */
 
-        crc0 = crc ^ 0xffffffff;
+    #define POLY 0x82f63b78
 
 
 
-        /* compute the crc for up to seven leading bytes to bring the data pointer
+    static void crc32c_word_table(void) {
 
-           to an eight-byte boundary */
+        uint32_t table[8][256];
 
-        while (len && ((uintptr_t)next & 7) != 0) {
 
-            __asm__("crc32b\t" "(%1), %0"
 
-                    : "=r"(crc0)
+        // Generate byte-wise table.
 
-                    : "r"(next), "0"(crc0));
+        for (unsigned n = 0; n < 256; n++) {
 
-            next++;
+            uint32_t crc = ~n;
+
+            for (unsigned k = 0; k < 8; k++)
 
-            len--;
+                crc = crc & 1 ? (crc >> 1) ^ POLY : crc >> 1;
 
+            table[0][n] = ~crc;
+
         }
 
 
 
-        /* compute the crc on sets of LONG*3 bytes, executing three independent crc
+        // Use byte-wise table to generate word-wise table.
 
-           instructions, each on LONG bytes -- this is optimized for the Nehalem,
+        for (unsigned n = 0; n < 256; n++) {
 
-           Westmere, Sandy Bridge, and Ivy Bridge architectures, which have a
+            uint32_t crc = ~table[0][n];
 
-           throughput of one crc per cycle, but a latency of three cycles */
+            for (unsigned k = 1; k < 8; k++) {
 
-        while (len >= LONG*3) {
+                crc = table[0][crc & 0xff] ^ (crc >> 8);
 
-            crc1 = 0;
+                table[k][n] = ~crc;
 
-            crc2 = 0;
+            }
 
-            end = next + LONG;
-
-            do {
-
-                __asm__("crc32q\t" "(%3), %0\n\t"
-
-                        "crc32q\t" LONGx1 "(%3), %1\n\t"
-
-                        "crc32q\t" LONGx2 "(%3), %2"
-
-                        : "=r"(crc0), "=r"(crc1), "=r"(crc2)
-
-                        : "r"(next), "0"(crc0), "1"(crc1), "2"(crc2));
-
-                next += 8;
-
-            } while (next < end);
-
-            crc0 = crc32c_shift(crc32c_long, crc0) ^ crc1;
-
-            crc0 = crc32c_shift(crc32c_long, crc0) ^ crc2;
-
-            next += LONG*2;
-
-            len -= LONG*3;
-
         }
 
 
 
-        /* do the same thing, but now on SHORT*3 blocks for the remaining data less
+        // Print table.
 
-           than a LONG*3 block */
+        print_table(table[0], 8, 256, "crc32c_table");
 
-        while (len >= SHORT*3) {
+    }
 
-            crc1 = 0;
 
-            crc2 = 0;
 
-            end = next + SHORT;
+    // Return a(x) multiplied by b(x) modulo p(x), where p(x) is the CRC
 
-            do {
+    // polynomial. For speed, this requires that a not be zero.
 
-                __asm__("crc32q\t" "(%3), %0\n\t"
+    static uint32_t multmodp(uint32_t a, uint32_t b) {
 
-                        "crc32q\t" SHORTx1 "(%3), %1\n\t"
+        uint32_t prod = 0;
 
-                        "crc32q\t" SHORTx2 "(%3), %2"
+        for (;;) {
 
-                        : "=r"(crc0), "=r"(crc1), "=r"(crc2)
+            if (a & 0x80000000) {
 
-                        : "r"(next), "0"(crc0), "1"(crc1), "2"(crc2));
+                prod ^= b;
 
-                next += 8;
+                if ((a & 0x7fffffff) == 0)
 
-            } while (next < end);
+                    break;
 
-            crc0 = crc32c_shift(crc32c_short, crc0) ^ crc1;
+            }
 
-            crc0 = crc32c_shift(crc32c_short, crc0) ^ crc2;
+            a <<= 1;
 
-            next += SHORT*2;
+            b = b & 1 ? (b >> 1) ^ POLY : b >> 1;
 
-            len -= SHORT*3;
-
         }
 
+        return prod;
 
+    }
 
-        /* compute the crc on the remaining eight-byte units less than a SHORT*3
 
-           block */
 
-        end = next + (len - (len & 7));
+    /* Take a length and build four lookup tables for applying the zeros operator
 
-        while (next < end) {
+       for that length, byte-by-byte, on the operand. */
 
-            __asm__("crc32q\t" "(%1), %0"
+    static void crc32c_zero_table(size_t len, char *name) {
 
-                    : "=r"(crc0)
+        // Generate operator for len zeros.
 
-                    : "r"(next), "0"(crc0));
+        uint32_t op = 0x80000000;               // 1 (x^0)
 
-            next += 8;
+        uint32_t sq = op >> 4;                  // x^4
 
-        }
-
-        len &= 7;
-
-
-
-        /* compute the crc for up to seven trailing bytes */
-
         while (len) {
 
-            __asm__("crc32b\t" "(%1), %0"
+            sq = multmodp(sq, sq);              // x^2^(k+3), k == len bit position
 
-                    : "=r"(crc0)
+            if (len & 1)
 
-                    : "r"(next), "0"(crc0));
+                op = multmodp(sq, op);
 
-            next++;
+            len >>= 1;
 
-            len--;
-
         }
 
 
 
-        /* return a post-processed crc */
+        // Generate table to update each byte of a CRC using op.
 
-        return (uint32_t)crc0 ^ 0xffffffff;
+        uint32_t table[4][256];
 
-    }
+        for (unsigned n = 0; n < 256; n++) {
 
+            table[0][n] = multmodp(op, n);
 
+            table[1][n] = multmodp(op, n << 8);
 
-    /* Check for SSE 4.2.  SSE 4.2 was first supported in Nehalem processors
+            table[2][n] = multmodp(op, n << 16);
 
-       introduced in November, 2008.  This does not check for the existence of the
+            table[3][n] = multmodp(op, n << 24);
 
-       cpuid instruction itself, which was introduced on the 486SL in 1992, so this
+        }
 
-       will fail on earlier x86 processors.  cpuid works on all Pentium and later
 
-       processors. */
 
-    #define SSE42(have) \
+        // Print the table to stdout.
 
-        do { \
+        print_table(table[0], 4, 256, name);
 
-            uint32_t eax, ecx; \
-
-            eax = 1; \
-
-            __asm__("cpuid" \
-
-                    : "=c"(ecx) \
-
-                    : "a"(eax) \
-
-                    : "%ebx", "%edx"); \
-
-            (have) = (ecx >> 20) & 1; \
-
-        } while (0)
-
-
-
-    /* Compute a CRC-32C.  If the crc32 instruction is available, use the hardware
-
-       version.  Otherwise, use the software version. */
-
-    uint32_t crc32c(uint32_t crc, const void *buf, size_t len)
-
-    {
-
-        int sse42;
-
-
-
-        SSE42(sse42);
-
-        return sse42 ? crc32c_hw(crc, buf, len) : crc32c_sw(crc, buf, len);
-
     }
 
 
 
-    #ifdef TEST
+    int main(void) {
 
+        puts(
 
+    "// crc32c.h\n"
 
-    #define SIZE (262144*3)
+    "// Tables and constants for crc32c.c software and hardware calculations.\n"
 
-    #define CHUNK SIZE
+    "\n"
 
+    "// Table for a 64-bits-at-a-time software CRC-32C calculation. This table\n"
 
+    "// has built into it the pre and post bit inversion of the CRC."
 
-    int main(int argc, char **argv)
+        );
 
-    {
+        crc32c_word_table();
 
-        char *buf;
+        puts(
 
-        ssize_t got;
+    "\n// Block sizes for three-way parallel crc computation.  LONG and SHORT\n"
 
-        size_t off, n;
+    "// must both be powers of two.  The associated string constants must be set\n"
 
-        uint32_t crc;
+    "// accordingly, for use in constructing the assembler instructions."
 
+            );
 
+        printf("#define LONG %d\n", LONG);
 
-        (void)argv;
+        printf("#define LONGx1 \"%d\"\n", LONG);
 
-        crc = 0;
+        printf("#define LONGx2 \"%d\"\n", 2 * LONG);
 
-        buf = malloc(SIZE);
+        printf("#define SHORT %d\n", SHORT);
 
-        if (buf == NULL) {
+        printf("#define SHORTx1 \"%d\"\n", SHORT);
 
-            fputs("out of memory", stderr);
+        printf("#define SHORTx2 \"%d\"\n", 2 * SHORT);
 
-            return 1;
+        puts(
 
-        }
+    "\n// Table to shift a CRC-32C by LONG bytes."
 
-        while ((got = read(0, buf, SIZE)) > 0) {
+        );
 
-            off = 0;
+        crc32c_zero_table(8192, "crc32c_long");
 
-            do {
+        puts(
 
-                n = (size_t)got - off;
+    "\n// Table to shift a CRC-32C by SHORT bytes."
 
-                if (n > CHUNK)
+        );
 
-                    n = CHUNK;
+        crc32c_zero_table(256, "crc32c_short");
 
-                crc = argc > 1 ? crc32c_sw(crc, buf + off, n) :
-
-                                 crc32c(crc, buf + off, n);
-
-                off += n;
-
-            } while (off < (size_t)got);
-
-        }
-
-        free(buf);
-
-        if (got == -1) {
-
-            fputs("read error\n", stderr);
-
-            return 1;
-
-        }
-
-        printf("%08x\n", crc);
-
         return 0;
 
     }
-
-
-
-    #endif /* TEST */
```

**Comment by stevengj** [comment]
Here is the `diff -u` of just the core computational function, `crc32_hw`:
```diff
--- oldcode.c	2023-11-27 18:12:19
+++ newcode.c	2023-11-27 18:12:35
@@ -1,40 +1,33 @@
-/* Compute CRC-32C using the Intel hardware instruction. */
-static uint32_t crc32c_hw(uint32_t crc, const void *buf, size_t len)
-{
-    const unsigned char *next = buf;
-    const unsigned char *end;
-    uint64_t crc0, crc1, crc2;      /* need to be 64 bits for crc32q */
+static uint32_t crc32c_hw(uint32_t crc, void const *buf, size_t len) {
+    if (buf == NULL)
+        return 0;
 
-    /* populate shift tables the first time through */
-    pthread_once(&crc32c_once_hw, crc32c_init_hw);
+    // Pre-process the crc.
+    uint64_t crc0 = crc ^ 0xffffffff;
 
-    /* pre-process the crc */
-    crc0 = crc ^ 0xffffffff;
-
-    /* compute the crc for up to seven leading bytes to bring the data pointer
-       to an eight-byte boundary */
+    // Compute the crc for up to seven leading bytes, bringing the data pointer
+    // to an eight-byte boundary.
+    unsigned char const *next = buf;
     while (len && ((uintptr_t)next & 7) != 0) {
         __asm__("crc32b\t" "(%1), %0"
-                : "=r"(crc0)
-                : "r"(next), "0"(crc0));
+                : "+r"(crc0)
+                : "r"(next), "m"(*next));
         next++;
         len--;
     }
 
-    /* compute the crc on sets of LONG*3 bytes, executing three independent crc
-       instructions, each on LONG bytes -- this is optimized for the Nehalem,
-       Westmere, Sandy Bridge, and Ivy Bridge architectures, which have a
-       throughput of one crc per cycle, but a latency of three cycles */
+    // Compute the crc on sets of LONG*3 bytes, making use of three ALUs in
+    // parallel on a single core.
     while (len >= LONG*3) {
-        crc1 = 0;
-        crc2 = 0;
-        end = next + LONG;
+        uint64_t crc1 = 0;
+        uint64_t crc2 = 0;
+        unsigned char const *end = next + LONG;
         do {
             __asm__("crc32q\t" "(%3), %0\n\t"
                     "crc32q\t" LONGx1 "(%3), %1\n\t"
                     "crc32q\t" LONGx2 "(%3), %2"
-                    : "=r"(crc0), "=r"(crc1), "=r"(crc2)
-                    : "r"(next), "0"(crc0), "1"(crc1), "2"(crc2));
+                    : "+r"(crc0), "+r"(crc1), "+r"(crc2)
+                    : "r"(next), "m"(*next));
             next += 8;
         } while (next < end);
         crc0 = crc32c_shift(crc32c_long, crc0) ^ crc1;
@@ -43,18 +36,18 @@
         len -= LONG*3;
     }
 
-    /* do the same thing, but now on SHORT*3 blocks for the remaining data less
-       than a LONG*3 block */
+    // Do the same thing, but now on SHORT*3 blocks for the remaining data less
+    // than a LONG*3 block.
     while (len >= SHORT*3) {
-        crc1 = 0;
-        crc2 = 0;
-        end = next + SHORT;
+        uint64_t crc1 = 0;
+        uint64_t crc2 = 0;
+        unsigned char const *end = next + SHORT;
         do {
             __asm__("crc32q\t" "(%3), %0\n\t"
                     "crc32q\t" SHORTx1 "(%3), %1\n\t"
                     "crc32q\t" SHORTx2 "(%3), %2"
-                    : "=r"(crc0), "=r"(crc1), "=r"(crc2)
-                    : "r"(next), "0"(crc0), "1"(crc1), "2"(crc2));
+                    : "+r"(crc0), "+r"(crc1), "+r"(crc2)
+                    : "r"(next), "m"(*next));
             next += 8;
         } while (next < end);
         crc0 = crc32c_shift(crc32c_short, crc0) ^ crc1;
@@ -63,26 +56,26 @@
         len -= SHORT*3;
     }
 
-    /* compute the crc on the remaining eight-byte units less than a SHORT*3
-       block */
-    end = next + (len - (len & 7));
+    // Compute the crc on the remaining eight-byte units less than a SHORT*3
+    // block.
+    unsigned char const *end = next + (len - (len & 7));
     while (next < end) {
         __asm__("crc32q\t" "(%1), %0"
-                : "=r"(crc0)
-                : "r"(next), "0"(crc0));
+                : "+r"(crc0)
+                : "r"(next), "m"(*next));
         next += 8;
     }
     len &= 7;
 
-    /* compute the crc for up to seven trailing bytes */
+    // Compute the crc for up to seven trailing bytes.
     while (len) {
         __asm__("crc32b\t" "(%1), %0"
-                : "=r"(crc0)
-                : "r"(next), "0"(crc0));
+                : "+r"(crc0)
+                : "r"(next), "m"(*next));
         next++;
         len--;
     }
 
-    /* return a post-processed crc */
-    return (uint32_t)crc0 ^ 0xffffffff;
+    // Return the crc, post-processed.
+    return ~(uint32_t)crc0;
 }
```

**Comment by stevengj** [comment]
(There were a bunch of other changes related to the generation of the tables, but we had already independently made our tables constant so I don't think we need those.   There were also stylistic changes that we don't need.)
